clippy's automatic fix for map_entry fails to compile, with a borrow error

### Summary

I've been doing Advent of Code to learn Rust and I wrote some code which involves counting the appearances of each character in a string. Clippy suggests that I can make the code more efficient by using `entry`, but the proposed fix fails due to some borrow errors. The proposed fix doesn't work, but I was able to use `entry` and the compiler's feedback to get some code which does compile.

### Reproducer

I tried this code:

```rust
let hand = "ABCDABCAA";
let mut counts: HashMap<char, u64> = HashMap::new();
for c in hand.chars() {
    if !counts.contains_key(&c) {
        counts.insert(c, 1);
    } else {
        counts.insert(c, counts.get(&c).unwrap() + 1);
    }
}
```

I ran `cargo clippy --fix` and got the following output:
```
warning: failed to automatically apply fixes suggested by rustc to crate `day_7`

after fixes were automatically applied the compiler reported errors within these files:

  * src/day_7.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see 
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0502]: cannot borrow `counts` as immutable because it is also borrowed as mutable
  --> src/day_7.rs:97:34
   |
92 |                 match counts.entry(c) {
   |                       ------ mutable borrow occurs here
...
97 |                         e.insert(counts.get(&c).unwrap() + 1);
   |                           ------ ^^^^^^ immutable borrow occurs here
   |                           |
   |                           mutable borrow later used by call

error[E0502]: cannot borrow `counts` as immutable because it is also borrowed as mutable
   --> src/day_7.rs:199:26
    |
194 |         match counts.entry(c) {
    |               ------ mutable borrow occurs here
...
199 |                 e.insert(counts.get(&c).unwrap() + 1);
    |                   ------ ^^^^^^ immutable borrow occurs here
    |                   |
    |                   mutable borrow later used by call

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0502`.
Original diagnostics will follow.

warning: function `silly` is never used
   --> src/day_7.rs:187:4
    |
187 | fn silly() {
    |    ^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: usage of `contains_key` followed by `insert` on a `HashMap`
  --> src/day_7.rs:92:17
   |
92 | /                 if !counts.contains_key(&c) {
93 | |                     counts.insert(c, 1);
94 | |                 } else {
95 | |                     counts.insert(c, counts.get(&c).unwrap() + 1);
96 | |                 }
   | |_________________^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_entry
   = note: `#[warn(clippy::map_entry)]` on by default
help: try
   |
92 ~                 match counts.entry(c) {
93 +                     std::collections::hash_map::Entry::Vacant(e) => {
94 +                         e.insert(1);
95 +                     }
96 +                     std::collections::hash_map::Entry::Occupied(mut e) => {
97 +                         e.insert(counts.get(&c).unwrap() + 1);
98 +                     }
99 +                 }
   |

warning: usage of `contains_key` followed by `insert` on a `HashMap`
   --> src/day_7.rs:191:9
    |
191 | /         if !counts.contains_key(&c) {
192 | |             counts.insert(c, 1);
193 | |         } else {
194 | |             counts.insert(c, counts.get(&c).unwrap() + 1);
195 | |         }
    | |_________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_entry
help: try
    |
191 ~         match counts.entry(c) {
192 +             std::collections::hash_map::Entry::Vacant(e) => {
193 +                 e.insert(1);
194 +             }
195 +             std::collections::hash_map::Entry::Occupied(mut e) => {
196 +                 e.insert(counts.get(&c).unwrap() + 1);
197 +             }
198 +         }
    |

warning: `all-days` (bin "day_7" test) generated 3 warnings (run `cargo clippy --fix --bin "day_7" --tests` to apply 2 suggestions)
warning: failed to automatically apply fixes suggested by rustc to crate `day_7`

after fixes were automatically applied the compiler reported errors within these files:

  * src/day_7.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see 
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0502]: cannot borrow `counts` as immutable because it is also borrowed as mutable
  --> src/day_7.rs:97:34
   |
92 |                 match counts.entry(c) {
   |                       ------ mutable borrow occurs here
...
97 |                         e.insert(counts.get(&c).unwrap() + 1);
   |                           ------ ^^^^^^ immutable borrow occurs here
   |                           |
   |                           mutable borrow later used by call

error[E0502]: cannot borrow `counts` as immutable because it is also borrowed as mutable
   --> src/day_7.rs:199:26
    |
194 |         match counts.entry(c) {
    |               ------ mutable borrow occurs here
...
199 |                 e.insert(counts.get(&c).unwrap() + 1);
    |                   ------ ^^^^^^ immutable borrow occurs here
    |                   |
    |                   mutable borrow later used by call

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0502`.
Original diagnostics will follow.

warning: `all-days` (bin "day_7") generated 3 warnings (3 duplicates)
```

### Version

```text
rustc 1.74.0 (79e9716c9 2023-11-13)
binary: rustc
commit-hash: 79e9716c980570bfd1f666e3b16ac583f0168962
commit-date: 2023-11-13
host: aarch64-apple-darwin
release: 1.74.0
LLVM version: 17.0.4
```


### Additional Labels

_No response_

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"Ethiraric"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/clippy_lints/src/entry.rs
+++ b/clippy_lints/src/entry.rs
@@ -214,12 +214,31 @@ impl MapType {
     }
 }
 
+/// Details on an expression checking whether a map contains a key.
+///
+/// For instance, with the following:
+/// ```ignore
+/// !!!self.the_map.contains_key("the_key")
+/// ```
+///
+/// - `negated` will be set to `true` (the 3 `!` negate the condition)
+/// - `map` will be the `self.the_map` expression
+/// - `key` will be the `"the_key"` expression
 struct ContainsExpr<'tcx> {
+    /// Whether the check for `contains_key` was negated.
     negated: bool,
+    /// The map on which the check is performed.
     map: &'tcx Expr<'tcx>,
+    /// The key that is checked to be contained.
     key: &'tcx Expr<'tcx>,
+    /// The context of the whole condition expression.
     call_ctxt: SyntaxContext,
 }
+
+/// Inspect the given expression and return details about the `contains_key` check.
+///
+/// If the given expression is not a `contains_key` check against a `BTreeMap` or a `HashMap`,
+/// return `None`.
 fn try_parse_contains<'tcx>(cx: &LateContext<'_>, expr: &'tcx Expr<'_>) -> Option<(MapType, ContainsExpr<'tcx>)> {
     let mut negated = false;
     let expr = peel_hir_expr_while(expr, |e| match e.kind {
@@ -229,6 +248,7 @@ fn try_parse_contains<'tcx>(cx: &LateContext<'_>, expr: &'tcx Expr<'_>) -> Optio
         },
         _ => None,
     });
+
     match expr.kind {
         ExprKind::MethodCall(
             _,
@@ -261,11 +281,28 @@ fn try_parse_contains<'tcx>(cx: &LateContext<'_>, expr: &'tcx Expr<'_>) -> Optio
     }
 }
 
+/// Details on an expression inserting a key into a map.
+///
+/// For instance, on the following:
+/// ```ignore
+/// self.the_map.insert("the_key", 3 + 4);
+/// ```
+///
+/// - `map` will be the `self.the_map` expression
+/// - `key` will be the `"the_key"` expression
+/// - `value` will be the `3 + 4` expression
 struct InsertExpr<'tcx> {
+    /// The map into which the insertion is performed.
     map: &'tcx Expr<'tcx>,
+    /// The key at which to insert.
     key: &'tcx Expr<'tcx>,
+    /// The value to insert.
     value: &'tcx Expr<'tcx>,
 }
+
+/// Inspect the given expression and return details about the `insert` call.
+///
+/// If the given expression is not an `insert` call into a `BTreeMap` or a `HashMap`, return `None`.
 fn try_parse_insert<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<InsertExpr<'tcx>> {
     if let ExprKind::MethodCall(_, map, [key, value], _) = expr.kind {
         let id = cx.typeck_results().type_dependent_def_id(expr.hir_id)?;
@@ -298,7 +335,7 @@ struct Insertion<'tcx> {
     value: &'tcx Expr<'tcx>,
 }
 
-/// This visitor needs to do a multiple things:
+/// This visitor needs to do multiple things:
 /// * Find all usages of the map. An insertion can only be made before any other usages of the map.
 /// * Determine if there's an insertion using the same key. There's no need for the entry api
 ///   otherwise.
@@ -346,13 +383,26 @@ impl<'tcx> InsertSearcher<'_, 'tcx> {
         res
     }
 
-    /// Visits an expression which is not itself in a tail position, but other sibling expressions
+    /// Visit an expression which is not itself in a tail position, but other sibling expressions
     /// may be. e.g. if conditions
     fn visit_non_tail_expr(&mut self, e: &'tcx Expr<'_>) {
         let in_tail_pos = self.in_tail_pos;
         self.in_tail_pos = false;
         self.visit_expr(e);
         self.in_tail_pos = in_tail_pos;
+    }
+
+    /// Visit the key and value expression of an insert expression.
+    /// There may not be uses of the map in either of those two either.
+    fn visit_insert_expr_arguments(&mut self, e: &InsertExpr<'tcx>) {
+        let in_tail_pos = self.in_tail_pos;
+        let allow_insert_closure = self.allow_insert_closure;
+        let is_single_insert = self.is_single_insert;
+        walk_expr(self, e.key);
+        walk_expr(self, e.value);
+        self.in_tail_pos = in_tail_pos;
+        self.allow_insert_closure = allow_insert_closure;
+        self.is_single_insert = is_single_insert;
     }
 }
 impl<'tcx> Visitor<'tcx> for InsertSearcher<'_, 'tcx> {
@@ -425,6 +475,7 @@ impl<'tcx> Visitor<'tcx> for InsertSearcher<'_, 'tcx> {
 
         match try_parse_insert(self.cx, expr) {
             Some(insert_expr) if SpanlessEq::new(self.cx).eq_expr(self.map, insert_expr.map) => {
+                self.visit_insert_expr_arguments(&insert_expr);
                 // Multiple inserts, inserts with a different key, and inserts from a macro can't use the entry api.
                 if self.is_map_used
                     || !SpanlessEq::new(self.cx).eq_expr(self.key, insert_expr.key)

--- a/tests/ui/entry.fixed
+++ b/tests/ui/entry.fixed
@@ -165,4 +165,15 @@ pub fn issue_10331() {
     }
 }
 
+/// Issue 11935
+/// Do not suggest using entries if the map is used inside the `insert` expression.
+pub fn issue_11935() {
+    let mut counts: HashMap<u64, u64> = HashMap::new();
+    if !counts.contains_key(&1) {
+        counts.insert(1, 1);
+    } else {
+        counts.insert(1, counts.get(&1).unwrap() + 1);
+    }
+}
+
 fn main() {}

--- a/tests/ui/entry.rs
+++ b/tests/ui/entry.rs
@@ -169,4 +169,15 @@ pub fn issue_10331() {
     }
 }
 
+/// Issue 11935
+/// Do not suggest using entries if the map is used inside the `insert` expression.
+pub fn issue_11935() {
+    let mut counts: HashMap<u64, u64> = HashMap::new();
+    if !counts.contains_key(&1) {
+        counts.insert(1, 1);
+    } else {
+        counts.insert(1, counts.get(&1).unwrap() + 1);
+    }
+}
+
 fn main() {}

